### DISCUSSION TITLE
Mixing ADKernel and Kernel

### DISCUSSION THREAD

**Discussion by rks171** [idaholab/moose#24111] [Q&A General] [answered]
I am testing out using the ADTimeKernel in my application in place of TimeKernel.  It is my understanding that it is ok to mix the AD classes and non-AD classes (all of the other kernels in my application are of the non-AD type currently).  After I do this, my test case stops converging.  What is confusing me, however, is that my Preconditioning block defines a "FDP" preconditioner.  I thought this meant the Jacobian was numerically calculated, so I was expecting the ADTimeKernel switch to have no impact until I switch to using the AD-calculated Jacobian.  Does the switch have some other impact on the solution?  Is the AD-calculated Jacobian still being used in my solution?

**Comment by lindsayad** [comment]
FDP should finite difference the residuals, so the implication here would be that the AD kernel's *residuals* are different than the non-AD kernels, which would definitely be concerning. Do you have an example input I could try?

**Comment by rks171** [comment, marked as answer]
Unfortunately, this is in a MOOSE application with local changes, so it's not very easy to share an input to reproduce the issue.  Maybe there is a problem with how I am switching out my `TimeKernel` for `ADTimeKernel`?  I inherit from `ADTimeKernel` instead of `TimeKernel`, comment out the `computeQpJacobian` and `computeQpOffDiagJacobian` methods, change the `computeQpResidual` method to type `ADReal`, and change all of my `Real` types to `ADReal` in the class.  My `computeQpResidual` method is very simple.  It is only returning `_rho_dot[_qp]*_test[_i][_qp]`.  Are there any other changes I would need to make when making the switch?

**Reply by GiudGiud** [reply]
How are you retrieving the dot variable value? Using couplesDot ?
There s an AD version of this to use

**Reply by rks171** [reply]
I am using `adCoupledDot`.

**Reply by GiudGiud** [reply]
can you share the code with us or is it restricted?
